dev/buildtool/spin_commands.py

```diff
@@ -84,15 +83,33 @@ def _do_can_skip_repository(self, repository):
     return all_exist
 
   def build_all_distributions(self, repository):
-    name = repository.name
+    spin_package_path = repository.origin
+    double_slash = spin_package_path.find('//')
+    # Trim spin_package_path to format go package path properly.
+    if spin_package_path.find('//') != -1:
+      spin_package_path = spin_package_path[double_slash+2:]
+    if spin_package_path[:-1] == '/':
+      spin_package_path = spin_package_path[:-1]
+
     source_info = self.source_code_manager.refresh_source_info(
       repository, self.options.build_number)
     self.__build_version = source_info.to_build_version()
-    config_root = repository.git_dir
 
-    check_subprocess('go get -d -v', cwd=config_root)
+    # NOTE: go build is opinionated about where the source repository lives --
+    # it expects the source repository to be under $GOPATH/src/<package path>,
+    # and will use that the source code in that directory *even if you clone
+    # your own copy of the source repo*.
+    #
+    # To facilitate go's pattern, we operate on the source under
+    # $GOPATH/src/<path to spin> to build spin.
+    gopath = check_subprocess('go env GOPATH')
```

Comment:
can you set the GOPATH to be the output directory for this command?
Pass env={'GOPATH': path} to the subprocess commands?
That way the command side effects are still encapsulated within this filesystem.

Reply:
I'll give it a try. I thought the first time I tried that I ran into trouble dynamically setting `GOPATH`, but I just walked through it again now successfully. Making changes and re-testing.

Reply:
Getting a lot of this during testing: 

```
package _/var/lib/jenkins/workspace/Build_PrimaryArtifacts/build_output/build_spin: unrecognized import path "_/var/lib/jenkins/workspace/Build_PrimaryArtifacts/build_output/build_spin" (import path does not begin with hostname)
```

Seems that go's assumptions pretty severely breaks the buildtool's `build_input -> build_output` pattern due to the package naming resolution.

Reply:
I'm going to revert this back to operating on go's opinions of where the source repositories live reading from `go env`.

Reply:
can I see what it looked like ? I dont see how that would matter.

You need input_path for the source analysis stuff (changelog, etc).
Since go is opinionated about where it goes, i guess you can make the GOPATH the commands build_output. Then when you run the command it will checkout again and build in the output dir. It breaks assumptions but is technically ok. The downside is an extra checkout and when we archive jenkins builds, we'll be archiving that source tree as well. So perhaps the GOPATH could be a different input directory (a subdir within the commands input dir).

The output directory is only used to direct log files and such. I dont know why your commands would even know about it. Perhaps you are executing in that directory. You can pass a different cwd to the input directory, but if you are setting GOPATH I'm not sure why go would be confused.

Reply:
I didn't see this before I merged so I assumed everything was OK. I'll follow up with another PR if I can clean any of this up. Sorry.